File: backend/realtime/live_readback.py
from __future__ import annotations


def search_live_blocks(blocks: list[dict[str, object]], query: str) -> list[dict[str, object]]:
    normalized = query.strip().lower()
    if not normalized:
        return blocks
    return [
        block
        for block in blocks
        if normalized in str(block.get("raw_text", "")).lower()
        or normalized in str(block.get("speaker_label", "")).lower()
    ]

File: frontend/assets/js/realtime/live_readback.js
window.liveReadback = {
    filter(blocks, query) {
        const normalized = String(query || '')
            .trim()
            .toLowerCase();
        if (!normalized) {
            return blocks;
        }
        return blocks.filter(
            (block) =>
                String(block.raw_text || '')
                    .toLowerCase()
                    .includes(normalized) ||
                String(block.speaker_label || '')
                    .toLowerCase()
                    .includes(normalized),
        );
    },
};
